### PR TITLE
Shipping Label Packages: Step 1

### PR DESCRIPTION
**Pull request by hichamboushaba**
Implements the first task and some parts of the second task of #3019, which is displaying the list of products when trying to edit packages of a shipping label, and displays a spinner for editing the selected package.

I felt a bit lost when working on this, and maybe there are some easier ways or better solutions than what I used, so please tell me about them 🙂.

I'll try to explain some decisions that I made here:

1. Loading packages is done in `CreateShippingLabelsViewModel` and not `EditShippingLabelPackagesViewModel`, I did this because I want to cache those packages, to avoid reloading them if the user wants to edit their changes later. I couldn't cache them in the `ShippingLabelRepository`, as I would have to make it a singleton, but this means we will cache them for the app lifecycle.
2. I combined both the custom packages, and predefined one in the same app model: `ShippingPackage`, this makes handling the list easier, and we can differentiate between them using the `category` field.
3. `ShippingPackage` has an `id` field, this is not used now, but we will need it to fetch the rates, and it has to be added in FluxC, I'll do it in the upcoming PRs.
4. The `EditShippingLabelPackagesFragment` may show a skeleton to fetch the order products if they weren't cached before, and in case of errors during fetching, we will just ignore the error and hide the weight, this was confirmed by Garance.

Both Unit tests, and the rest of the task will be done in upcoming PRs, as this is already too big.

#### Testing

- Prepare an order on the web that satisfies SL creation preconditions.
- Go to Orders -> Open the order
- Pass the steps of origin and destination addresses.
- Click on Continue of Package details.
- Notice that a loading dialog appears, and then the package details screen will be displayed.
- Go Back.
- Click on Continue a second time.
- Notice the package details screen is opened without reloading packages.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by hichamboushaba** [comment]
@0nko please note that testing this PR is impacted by the crash #3511, so to test it, you need to merge the changes from this PR #3515 locally.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+fun CustomPackage.toAppModel(): ShippingPackage {
+    val dimensionsParts = dimensions.split("x")
+    return ShippingPackage(
+        title = title,
+        isLetter = isLetter,
+        dimensions = PackageDimensions(
+            length = dimensionsParts[0].trim().toDouble(),
+            width = dimensionsParts[0].trim().toDouble(),
+            height = dimensionsParts[0].trim().toDouble()
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt]: I think the indexes above should be 1 and 2.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+fun PredefinedOption.toAppModel(): List<ShippingPackage> {
+    return predefinedPackages.map {
+        val dimensionsParts = it.dimensions.split("x")
+        ShippingPackage(
+            title = it.title,
+            isLetter = it.isLetter,
+            dimensions = PackageDimensions(
+                length = dimensionsParts[0].trim().toDouble(),
+                width = dimensionsParts[0].trim().toDouble(),
+                height = dimensionsParts[0].trim().toDouble()
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt]: Same here.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt [hunk 3]
```diff
@@ -103,9 +115,39 @@ class CreateShippingLabelViewModel @AssistedInject constructor(
     }
 
     private suspend fun handleResult(action: suspend () -> Event) {
-        viewState = viewState.copy(isProgressDialogVisible = true)
+        val progressDialogState = ProgressDialogState(
+            isShown = true,
+            title = string.shipping_label_edit_address_validation_progress_title,
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt]: This method is also called by the `LoadData` side effect, so the progress dialog message should be different. 

I know it was like that before, I just noticed it after your refactoring.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesFragment.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    fun showSkeleton(show: Boolean, binding: FragmentEditShippingLabelPackagesBinding) {
+        if (show) {
+            skeletonView.show(binding.packagesList, R.layout.skeleton_shipping_label_package_details, delayed = false)
+        } else {
+            skeletonView.hide()
+        }
+    }
+
+    override fun onRequestAllowBackPress(): Boolean {
+        navigateBackWithNotice(EDIT_PACKAGES_CLOSED)
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesFragment.kt:76]: Navigating back directly from the fragment bypasses the ViewModel, so we won't be able to unit test the screen exit event.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+class EditShippingLabelPackagesViewModel @AssistedInject constructor(
+    @Assisted savedState: SavedStateWithArgs,
+    dispatchers: CoroutineDispatchers,
+    parameterRepository: ParameterRepository,
+    private val orderDetailRepository: OrderDetailRepository,
+    private val productDetailRepository: ProductDetailRepository
+) : ScopedViewModel(savedState, dispatchers) {
+    companion object {
+        private const val KEY_PARAMETERS = "ket_parameters"
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt]: NP: Typo.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelPackagesAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+import com.woocommerce.android.ui.orders.shippinglabels.creation.ShippingLabelPackagesAdapter.ShippingLabelPackageViewHolder
+import com.woocommerce.android.ui.products.models.SiteParameters
+import com.woocommerce.android.util.StringUtils
+
+class ShippingLabelPackagesAdapter(
+    val parameters: SiteParameters,
+    val onWeightEdited: (Int, Double) -> Unit
+) : RecyclerView.Adapter<ShippingLabelPackageViewHolder>() {
+    var shippingLabelPackages: List<ShippingLabelPackage> = emptyList()
+        set(value) {
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelPackagesAdapter.kt:25]: Nice 👍

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelPackagesAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+        @SuppressLint("SetTextI18n")
+        fun bind(position: Int) {
+            val context = binding.root.context
+            val shippingLabelPackage = shippingLabelPackages[position]
+            binding.packageName.text = context.getString(
+                R.string.shipping_label_package_details_title_template,
+                position + 1
+            )
+            binding.packageItemsCount.text = " - ${context.getString(
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelPackagesAdapter.kt]: We could use the plural string resource to cover both 1 and more items.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        if (viewState.shippingLabelPackages.isNotEmpty()) return
+        launch {
+            loadProductsWeightsIfNeeded()
+            val packagesList = arguments.shippingLabelPackages.toList().ifEmpty {
+                val order = requireNotNull(orderDetailRepository.getOrder(arguments.orderId))
+                listOf(
+                    ShippingLabelPackage(
+                        selectedPackage = arguments.availablePackages.first(),
+                        weight = Double.NaN,
+                        items = order.items.map { it.toShippingItem() }
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt]: This shows all products for the order, including virtual and refunded products, which should be filtered out.

File: WooCommerce/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -387,7 +387,17 @@
...
     <string name="shipping_label_address_suggestion_suggested_address">Suggested address</string>
-
+    <string name="shipping_label_package_details_items_section_title">Items to fulfill</string>
+    <string name="shipping_label_package_details_move_item">Move</string>
+    <string name="shipping_label_package_details_section_title">Package details</string>
+    <string name="shipping_label_package_details_selected_package_hint">Package selected</string>
+    <string name="shipping_label_package_details_weight_hint">Total package weight (%1$s)</string>
+    <string name="shipping_label_package_details_weight_info">Includes package weight</string>
+    <string name="shipping_label_packages_loading_error">Unable to load package definitions</string>
+    <string name="shipping_label_package_details_items_count">"%1$d items  "</string>
```
Comment (0nko) [inline WooCommerce/src/main/res/values/strings.xml]: Is the extra whitespace intentional?

**Review by 0nko — COMMENTED** [review]
I've tested the PR and things look good 👍. 

I've left a couple of minor comments in the code. The biggest issue is that the packages show all order items, including refunds and virtual products.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt]: Do we need to show a progress dialog for LoadData side effect, for now, it's synchronous, and just loads the cached order?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesFragment.kt (hunk 4 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesFragment.kt:76]: Yes, this is just temporary, I was planning to update it in the upcoming PR, when I handle the `done` button click too.

File: WooCommerce/src/main/res/values/strings.xml (hunk 9 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/res/values/strings.xml]: Fixed in the last commit.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt (hunk 8 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt]: I added filtering in the last commit, I'm still ignoring errors if the products API call fails, I prefer to keep it this way in this PR, and confirm with Garance later if we should display an error message, and implement it in a separate PR.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelPackagesAdapter.kt (hunk 7 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelPackagesAdapter.kt]: Done.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt (hunk 1 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt]: Fixed in the last commit.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt (hunk 3 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt]: Probably not.

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt]: @0nko I updated handleResult to accept an optional prodress dialog title and message parameters in 40232a8d8a446bcdc817f1b9c7f4442edc92c47c, this way loadData won't display any dialog progress now, but if we need to load any async data later, we can add it, and we can reuse handleResult for other scenarios. WDYT?

Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt]: Sounds good!

**Review by 0nko — APPROVED** [review]
Nice work, Hicham! 🚀

**Pull request merged**